### PR TITLE
TICL Reorganisation

### PR DESCRIPTION
**Pull request by rovere**
##### TICL Reorganization

Several changes entered this PR.
    
  * The data structures used by TICL have been moved into
      DataFormats/TICL, that has to be added as a CMSSW package.
      Tracksters and Tile are currently present. The encapsulation of
      std::array as a transient data member of a Tile is due to the fact
      that ROOT natively supports those and refuses to create ad-hoc
      dictionaries. Moreover, at present, an std::array cannot sit at the
      top of a ROOT branch requiring, once again, its encapsulation.
  * The "master Tile" has been added as a single producer. A unique Tile
      containing all the 2D reconstructed clusters is created once and put
      into the Event. All iterations will fetch it from the event and mask
      the 2D clusters on the fly. This required also some changes to the
      clusters filters logic. Now the filters produce also a vector<float>
      that represents the result of merging their input mask with their
      own selection. TICL iterations will make use of the filter's
      vector<float> to dynamically mask clusters from the "master Tile".
  * The fractions have been added to the Tracksters. The sharing of the
      energy of 2D clusters is flat and based on their multiplicity, i.e,
      a fraction of energy/multiplicity is assigned to all Tracksters that
      use the 2D clusters. Better splitting algorithms can be studies or
      Machine Learning techniques could be put at work here.
  * The hits and fraction of the Trackster are now correctly populated.
      This required an additional loop over the components of each 2D
      cluster. In the future the CaloCluster, PFCluster and hits and
      fractions data structures could be reviewed or, at least, TICL could
      use some other, more efficient, internal data representation and
      collapse into the legacy ones only at the very last stage.

#### PR validation:

`runTheMatrix.py -l limited -i all -j 4`

and other samples.

**Note to the reviewers**: all changes in `RecoLocalCalo/HGCalRecAlgos` are purely due to the code-format. No other changes have been committed there. I did so so that PR #26991 could be closed (if and when this PR gets merged)

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@rovere 
your PR is missing clang-format changes in the following files

- RecoHGCal/TICL/plugins/HGCDoublet.cc
- RecoHGCal/TICL/plugins/HGCDoublet.h
- RecoLocalCalo/HGCalRecAlgos/src/RecHitTools.cc

Were these just missed in the commit file list or did your formatting tool somehow skipped these?
Please check and clarify.
Thank you.

File: RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h [hunk 1]
```diff
@@ -18,72 +18,67 @@
...
     }
-    
+
     DetId baseid = uncalibRH.id();
     unsigned layer = 0;
-    if( DetId::Hcal == baseid.det() && HcalEndcap == baseid.subdetId() ) {
-      layer =  HcalDetId(baseid).depth() + 40;
-    } else if ( DetId::Forward == baseid.det() && HGCEE == baseid.subdetId() ) {
+    if (DetId::Hcal == baseid.det() && HcalEndcap == baseid.subdetId()) {
+      layer = HcalDetId(baseid).depth() + 40;
```
Comment (kpedro88) [inline RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h]: these hardcoded values should come from the geometry or RecHitTools

**Comment by kpedro88** [comment]
This is an error from the IB, it should be fixed in the next IB.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h (hunk 1 above)
Comment (rovere) [reply RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h]: Ciao @kpedro88,
as I said in the comment to this PR the package `RecoLocalCalo/HGCCalRecAlgos` has been added only to be able to include all `code-format` changes that were in #26991 
As such, I'd not consider this package as part of this review.

**Comment by rovere** [comment]
@slava77 I have no private tool at hand to perform code formatting.
I run
`scram b code-format`
and I hope this will touch and format all files in a package, not just the ones that have been modified. 
In any case, thanks for spotting this: I can manually run the clang formatter on those files.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h (hunk 1 above)
Comment (kpedro88) [reply RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h]: Okay, I've added this to #25472.

**Comment by kpedro88** [comment]
@rovere there is a conflict, please rebase

**Comment by rovere** [comment]
This is **really** disturbing. I guess that Sunanda's changes were made w/o formatting the code properly.
The changes in `RecoLocalCalo`, let me stress it once again, are not naturally part of this PR.
If the rebase turn out to be non-trivial, I'd simply remove all the changes there and leave the formatting for a later PR that someone else will make.
I'm not sure why code-formatting has not been requested on #27015 while it has been here.

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
>  I'm not sure why code-formatting has not been requested on #27015 while it has been here.

the formatting (clang-format) is not required yet for the PRs to be tested.
It was not required for this PR either.

**Comment by rovere** [comment]
@slava77  by the time #27015 has been merged, this PR with code-formatting in the packages touched by #27015 was already open.

**Comment by fabiocos** [comment]
@rovere the clang-format option was not yet enabled in code checks waiting for the bulk of the migration to be done, to minimize side effects. Unfortunately the more we advance the less this is true. 
Now the bulk migration is almost over, and we will start forcing the option in the code-checks.

File: RecoHGCal/TICL/python/ticl_iterations.py [hunk 2]
```diff
@@ -19,6 +20,8 @@
 def TICL_iterations_withReco(process):
```
Comment (slava77) [inline RecoHGCal/TICL/python/ticl_iterations.py]: this file should eventually be replaced by parts that do not know about a process:
- a file with tasks defined to do the task
- a RecoHGCal/Configuration/python/RECOHGCal_EventContent_cff.py with all appropriate keep statements

**Review by slava77 — COMMENTED** [review]
After looking around at the available DataFromats/ packages organization,
I think that the proposed content of DataFormats/TICL should be moved to DataFormats/HGCRecHit.
The Trackster as a DAG is a bit more generic and could perhaps be eventually moved to some more common place when it is used outside of the TICL.

Since we are moving things around, I'd like to question the location of the TICL algorithms. It is added in RecoHGCal as the only package in that subsystem. I missed this somehow.
 The algorithms introduced so far as a part of TICL are local to HGCAL (everything that they produce are parts of HGC clustering). The TICL package better be moved to the RecoLocalCalo subsystem.

**Comment by slava77** [comment]
uhm, it looks like something else from RecoLocalCalo/HGCalRecAlgos was merged (was it #26996 ?).
Marco, you will need to rebase, sorry.
It was a nice idea to put the code formatting in this PR, but it apparently coincided with a lot of other HGCAL reco changes in other PRs.

**Comment by slava77** [comment]
If I'm not mistaken, we currently have #27049, #27128 and #27129 overlapping with this PR (in addition to three more that were already merged in the last few days; two rebases already made and one more expected).

**Comment by fabiocos** [comment]
This is what I can see:
```
Testing PR #  27026  for branch  master
File  RecoLocalCalo/HGCalRecAlgos/interface/HGCal3DClustering.h  modified in PRs #  ['27049', '27128', '27026']
File  RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h  modified in PRs #  ['27129', '27026']
File  RecoLocalCalo/HGCalRecAlgos/interface/KDTreeLinkerAlgoT.h  modified in PRs #  ['27049', '27026']
File  RecoLocalCalo/HGCalRecAlgos/interface/KDTreeLinkerToolsT.h  modified in PRs #  ['27049', '27026']
File  RecoLocalCalo/HGCalRecAlgos/interface/RecHitTools.h  modified in PRs #  ['27128', '27026']
File  RecoLocalCalo/HGCalRecAlgos/src/HGCal3DClustering.cc  modified in PRs #  ['27049', '27026']
File  RecoLocalCalo/HGCalRecAlgos/src/RecHitTools.cc  modified in PRs #  ['27128', '27026']
```

**Pull request closed without merging**